File: ShootR/ShootR/Server/Ship/ShipMovementController.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
namespace ShootR
{
    /// <summary>
    /// Ship's Movement Controller.  Handles all of the movement.
    /// </summary>
    public class ShipMovementController : MovementController
    {
        public const double ROTATE_SPEED = 180;
        public const double MASS = 50;
        public const double ENGINE_POWER = 110000;
        public const double DRAG_COEFFICIENT = .2;
        public const double DRAG_AREA = 5;

        private Vector2 _acceleration = Vector2.Zero;

        public ShipMovementController(Vector2 position) :
            base(MASS, ENGINE_POWER)
        {
            Position = position;
            Moving = new MovementFlags();
        }

        public MovementFlags Moving { get; set; }

        /// <summary>
        /// Calculates how far to move based on the current movement.
        /// </summary>
        /// <param name="percentOfSecond">How much time, percentage wise, has passed since the last Update.</param>
        public override void Move(double percentOfSecond)
        {
            double velocityLength;

            _acceleration = Forces / Mass;

            Position += Velocity * percentOfSecond + _acceleration * percentOfSecond * percentOfSecond;
            Velocity += _acceleration * percentOfSecond;
            velocityLength = Velocity.Length();

            // Stop moving if the "speed" is less than 10
            if (velocityLength < 10)
            {
                Velocity = Vector2.Zero;
            }
            else if (velocityLength > 3000) // Hack
            {
                Velocity = new Vector2(Rotation) * 600;
            }

            _acceleration = new Vector2();
            Forces = new Vector2();

            Vector2 direction = new Vector2(Rotation),
                    dragForce = .5 * Velocity * Velocity.Abs() * DRAG_COEFFICIENT * DRAG_AREA * -1;

            if (Moving.Forward)
            {
                ApplyForce(direction * Power);
            }
            if (Moving.Backward)
            {
                ApplyForce(direction * Power * -1);
            }

            ApplyForce(dragForce);

            double rotationIncrementor = percentOfSecond * ROTATE_SPEED;

            if (Moving.RotatingLeft)
            {
                Rotation -= rotationIncrementor;
            }
            if (Moving.RotatingRight)
            {
                Rotation += rotationIncrementor;
            }
        }

        public void StopMovement()
        {
            FlagMovement(Movement.Forward, false);
            FlagMovement(Movement.Backward, false);
            FlagMovement(Movement.RotatingLeft, false);
            FlagMovement(Movement.RotatingRight, false);
        }

        /// <summary>
        /// Used to register a start or a stop of a movement in a given direction.
        /// </summary>
        /// <param name="where">Where the Ship is moving.</param>
        /// <param name="flag">Whether to start or stop moving in the <paramref name="where"/> direction.</param>
        private void FlagMovement(Movement where, bool flag)
        {
            if (where == Movement.RotatingLeft)
            {
                Moving.RotatingLeft = flag;
            }
            else if (where == Movement.RotatingRight)
            {
                Moving.RotatingRight = flag;
            }
            else if (where == Movement.Forward)
            {
                Moving.Forward = flag;
            }
            else if (where == Movement.Backward)
            {
                Moving.Backward = flag;
            }
        }

        /// <summary>
        /// Starts the movement in the given direction <paramref name="where"/>.
        /// </summary>
        /// <param name="where">Where to start moving</param>
        public void StartMoving(Movement where)
        {
            FlagMovement(where, true);
        }

        /// <summary>
        /// Stops the movement in the given direction <paramref name="where"/>.
        /// </summary>
        /// <param name="where">Where to start moving</param>
        public void StopMoving(Movement where)
        {
            FlagMovement(where, false);
        }

        public void Update(GameTime gameTime)
        {
            Update(gameTime.PercentOfSecond);
        }

        public override void Update(double PercentOfSecond)
        {
            base.Update(PercentOfSecond);
            Move(PercentOfSecond);
        }
    }
}

File: ShootR/ShootR/Client2/Ships/ShipMovementController.ts
/// <reference path="../../Scripts/endgate-0.2.0-beta1.d.ts" />
/// <reference path="../Server/IPayloadDefinitions.ts" />
/// <reference path="IMoving.ts" />

module ShootR {

    export class ShipMovementController extends eg.MovementControllers.MovementController {
        public static MASS: number = 50;
        public static ENGINE_POWER: number = 110000;
        public static DRAG_AREA: number = 5;
        public static DRAG_COEFFICIENT: number = .2;
        public static ROTATE_SPEED: number = Math.PI;
        public static MOVING_DIRECTIONS: string[] = ["RotatingLeft", "RotatingRight", "Forward", "Backward"];
        public static INTERPOLATE_POSITION_THRESHOLD: number = 10;
        public static INTERPOLATE_ROTATION_THRESHOLD: number = 15;

        public Moving: IMoving;
        public Mass: number;
        public Power: number;
        public Forces: eg.Vector2d;
        private _acceleration: eg.Vector2d;        
        private _moveables: Array<eg.IMoveable>;

        constructor(movables: Array<eg.IMoveable>) {
            super(movables);

            this._moveables = movables;

            this.Mass = ShipMovementController.MASS;
            this.Power = ShipMovementController.ENGINE_POWER;
            this.Forces = eg.Vector2d.Zero;
            this._acceleration = eg.Vector2d.Zero;

            this.Moving = {
                Forward: false,
                Backward: false,
                RotatingLeft: false,
                RotatingRight: false
            };

            this.OnMove = new eg.EventHandler1<eg.MovementControllers.IMoveEvent>();
        }

        public OnMove: eg.EventHandler1<eg.MovementControllers.IMoveEvent>;

        public LoadPayload(payload: Server.IShipMovementControllerData): void {
            this.Rotation = payload.Rotation * .0174532925;
            this.Mass = payload.Mass;
            this.Forces = payload.Forces;
            this.Velocity = payload.Velocity;
            this.Position = payload.Position.Add(Ship.SIZE.Multiply(.5));
            this.Moving = payload.Moving;
        }

        public IsMovingInDirection(direction: string): boolean {
            return this.Moving[direction] || false;
        }

        public StartMoving(direction: string): void {
            this.Move(direction, true);
        }

        public StopMoving(direction: string): void {
            this.Move(direction, false);
        }

        public StopAllMovement(): void {
            for (var i = ShipMovementController.MOVING_DIRECTIONS.length - 1; i >= 0; i--) {
                this.Moving[ShipMovementController.MOVING_DIRECTIONS[i]] = false;
            }
        }

        public ApplyForce(force: eg.Vector2d): void {
            this.Forces = this.Forces.Add(force);
        }

        public Update(gameTime: eg.GameTime): void {
            var clientPositionPrediction: eg.Vector2d,
                rotationIncrementor: number,
                direction: eg.Vector2d = new eg.Vector2d(Math.cos(this.Rotation), Math.sin(this.Rotation)),
                dragForce: eg.Vector2d,
                velocityLength: number;

            this._acceleration = this.Forces.Divide(this.Mass);

            clientPositionPrediction = this.Velocity.Multiply(gameTime.Elapsed.Seconds).Add(this._acceleration.Multiply(gameTime.Elapsed.Seconds * gameTime.Elapsed.Seconds));
            this.Position = this.Position.Add(clientPositionPrediction);

            this.Velocity = this.Velocity.Add(this._acceleration.Multiply(gameTime.Elapsed.Seconds));

            velocityLength = this.Velocity.Length();

            // Stop moving if the "speed" is less than 10
            if (velocityLength < 10) {
                this.Velocity = eg.Vector2d.Zero;
            } else if (velocityLength > 3000) // Hack
            {
                this.Velocity = direction.Multiply(600);
            }

            this._acceleration = eg.Vector2d.Zero;
            this.Forces = eg.Vector2d.Zero;

            rotationIncrementor = gameTime.Elapsed.Seconds * ShipMovementController.ROTATE_SPEED;
            dragForce = this.Velocity.Multiply(.5).Multiply(this.Velocity.Abs()).Multiply(ShipMovementController.DRAG_COEFFICIENT * ShipMovementController.DRAG_AREA * -1);

            if (this.Moving.RotatingLeft) {
                this.Rotation -= rotationIncrementor;
            }
            if (this.Moving.RotatingRight) {
                this.Rotation += rotationIncrementor;
            }
            if (this.Moving.Forward) {
                this.ApplyForce(direction.Multiply(this.Power));
            }
            if (this.Moving.Backward) {
                this.ApplyForce(direction.Multiply(this.Power * -1));
            }

            this.ApplyForce(dragForce);

            for (var i = 0; i < this._moveables.length; i++) {
                this._moveables[i].Position = this.Position;

                if (this._moveables[i] instanceof eg.Graphics.Graphic2d) {
                    this._moveables[i].Rotation = this.Rotation;
                }
            }
        }

        public Move(direction: string, startMoving: boolean): void {
            if (typeof this.Moving[direction] !== "undefined") {
                this.Moving[direction] = startMoving;
                this.OnMove.Trigger(<eg.MovementControllers.IMoveEvent>{
                    Direction: direction,
                    StartMoving: startMoving
                });
            }
            else {
                throw new Error(direction + " is an unknown direction.");
            }
        }
    }

}